presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestOptimizedPartitionedOutputOperatorQueries.java

```diff
@@ -0,0 +1,560 @@
+                "    l.l_partkey = p.partkey";
+
+        Object expected = computeActual(noAriaSession(), sql).getOnlyValue();
+        Object result = computeActual(ariaSession(), sql).getOnlyValue();
+        assertEquals(result, expected);
+    }
+
+    @Test
+    public void testVarchar()
+    {
+        testVarchar("tpch");
+    }
+
+    @Test
+    public void testVarcharBucketed()
+    {
+        testVarchar("tpch_bucketed");
+    }
+
+    private void testVarchar(String schema)
+    {
+        String sql = "SELECT\n" +
+                "    checksum(l.comment)\n" +
+                "    FROM " + schema + ".lineitem l, " + schema + ".partsupp p\n" +
+                "WHERE\n" +
+                "    l.partkey = p.partkey\n" +
+                "    AND l.suppkey = p.suppkey\n" +
+                "    AND p.availqty < 1000";
+
+        Object expected = computeActual(noAriaSession(), sql).getOnlyValue();
+        Object result = computeActual(ariaSession(), sql).getOnlyValue();
+        assertEquals(result, expected);
+    }
+
+    @Test
+    public void testDictionaryOfVarcharWithNulls()
+    {
+        testDictionaryOfVarcharWithNulls("tpch");
+    }
+
+    @Test
+    public void testDictionaryOfVarcharWithNullsBucketed()
+    {
+        testDictionaryOfVarcharWithNulls("tpch_bucketed");
+    }
+
+    private void testDictionaryOfVarcharWithNulls(String schema)
+    {
+        String sql = "SELECT\n" +
+                "    checksum(l.shipmode)\n" +
+                "    FROM " + schema + ".lineitem l, " + schema + ".partsupp p\n" +
+                "WHERE\n" +
+                "    l.partkey = p.partkey\n" +
+                "    AND l.suppkey = p.suppkey\n" +
+                "    AND p.availqty < 1000";
+
+        Object expected = computeActual(noAriaSession(), sql).getOnlyValue();
+        Object result = computeActual(ariaSession(), sql).getOnlyValue();
+        assertEquals(result, expected);
+    }
+
+    @Test
+    public void testArrayOfBigInt()
+    {
+        testArrayOfBigInt("tpch");
+    }
+
+    @Test
+    public void testArrayOfBigIntBucketed()
+    {
+        testArrayOfBigInt("tpch_bucketed");
+    }
+
+    private void testArrayOfBigInt(String schema)
+    {
+        String sql = "WITH lineitem_array AS (\n" +
+                "    SELECT\n" +
+                "        partkey,\n" +
+                "        suppkey,\n" +
+                "        ARRAY[orderkey, partkey, suppkey] AS l_array\n" +
+                "    FROM " + schema + ".lineitem\n" +
+                ")\n" +
+                "SELECT\n" +
+                "    checksum(l.l_array)\n" +
+                "FROM lineitem_array l,\n" +
+                "    " + schema + ".partsupp p\n" +
+                "WHERE\n" +
+                "    l.partkey = p.partkey\n" +
+                "    AND l.suppkey = p.suppkey\n" +
+                "    AND p.availqty < 1000";
+
+        Object expected = computeActual(noAriaSession(), sql).getOnlyValue();
```

Comment:
Testing optimized re-partitioning using non-optimized repartitioning is problematic. When optimized version becomes stable enough, you'll want to delete non-optimized version. If you do that, you'll lose a bunch of test coverage. 

H2 has limited support for complex types, which makes it tricky to test them. However, in many cases it is still possible to use it. Here is how I'd re-write this test:

```
    private void testArrayOfBigInt(String schema)
    {
        String withClause = "WITH lineitem_ex AS (\n" +
                "    SELECT\n" +
                "        partkey,\n" +
                "        suppkey,\n" +
                "        ARRAY[orderkey, partkey, suppkey] AS keys\n" +
                "    FROM lineitem\n" +
                ")\n";

        String h2WithClause = "WITH lineitem_ex AS (\n" +
                "    SELECT\n" +
                "        partkey,\n" +
                "        suppkey,\n" +
                "        (orderkey, partkey, suppkey) AS keys\n" +
                "    FROM lineitem\n" +
                ")\n";

        String query = "SELECT keys FROM lineitem_ex l, partsupp p\n" +
                "WHERE\n" +
                "    l.partkey = p.partkey\n" +
                "    AND l.suppkey = p.suppkey\n" +
                "    AND p.availqty < 1000";

        Session session = Session.builder(getQueryRunner().getDefaultSession())
                .setSchema(schema)
                .build();
        assertQuery(session, withClause + query, h2WithClause + query);
    }
```

Another option could be to hard-code query results in file(s) and use that data for verification. See presto-product-tests/src/main/resources/sql-tests/testcases/hive_tpch folder.